pydra/engine/task.py

```diff
@@ -314,56 +314,159 @@ def command_args(self):
+        """Returining position and value for executable ShellTask input"""
+        pos = 0  # executable should be the first el. of the command
+        value = self._field_value(field=field, state_ind=state_ind, ind=ind)
+        if value is None:
+            raise Exception("executable has to be set")
+        return pos, ensure_list(value, tuple2list=True)
+
+    def _command_shelltask_args(self, field, state_ind, ind):
+        """Returining position and value for args ShellTask input"""
+        pos = -1  # assuming that args is the last el. of the command
+        value = self._field_value(
+            field=field, state_ind=state_ind, ind=ind, check_file=True
+        )
+        if value is None:
+            return None
+        else:
+            return pos, ensure_list(value, tuple2list=True)
+
+    def _command_pos_args(self, field, state_ind, ind):
+        """
+        Checking all additional input fields, setting pos to None, if position not set.
+        Creating a list with additional parts of the command that comes from
+        the specific field.
+        """
+        argstr = field.metadata.get("argstr", None)
+        if argstr is None:
+            if "output_file_template" in field.metadata:
+                # assuming that input that has output_file_template and no arstr
+                # are not used in the command
+                return None
             else:
-                cmd_add += ensure_list(value, tuple2list=True)
-            if cmd_add is not None:
-                pos_args.append((pos, cmd_add))
-        # sorting all elements of the command
-        pos_args.sort()
-        # if args available, they should be moved at the of the list
-        if pos_args[0][0] == -1:
-            pos_args.append(pos_args.pop(0))
-        # dropping the position index
-        cmd_args = []
-        for el in pos_args:
-            cmd_args += el[1]
-        return cmd_args
+                raise Exception(
+                    f"{field.name} doesn't have argstr field in the metadata"
+                )
+        pos = field.metadata.get("position", None)
+        if pos is None:
+            # position will be set at the end
+            pass
+        elif not isinstance(pos, int):
+            raise Exception(f"position should be an integer, but {pos} given")
+        elif pos == 0:
+            raise Exception(f"position can't be 0")
+        elif pos < 0:  # position -1 is for args
+            pos = pos - 1
+        # checking if the position is not already used
+        elif pos in self._positions_provided:
+            raise Exception(
+                f"{field.name} can't have provided position, {pos} is already used"
+            )
+        self._positions_provided.append(pos)
+        value = self._field_value(
+            field=field, state_ind=state_ind, ind=ind, check_file=True
+        )
+        if field.metadata.get("readonly", False) and value is not None:
```

Comment:
just a note that this may be the kind of metadata we check for on input assignment rather than just on commandline formation.

Reply:
perhaps it can be a part of input validation, once we build one

Reply:
at least basic input validation would be a good target for the august release (0.8).